tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/TableauEncoder.scala

```diff
@@ -0,0 +1,424 @@
+package at.forsyte.apalache.tla.pp.temporal
+
+import at.forsyte.apalache.tla.lir._
+
+import javax.inject.Singleton
+import com.typesafe.scalalogging.LazyLogging
+import scalaz.Scalaz.{init => _}
+import at.forsyte.apalache.tla.pp.UniqueNameGenerator
+import at.forsyte.apalache.tla.pp.IrrecoverablePreprocessingError
+import at.forsyte.apalache.tla.pp.temporal.utils.builder
+import at.forsyte.apalache.tla.pp.temporal.ScopedBuilderExtensions._
+import at.forsyte.apalache.tla.pp.temporal.DeclUtils._
+import scala.collection.mutable.HashMap
+import at.forsyte.apalache.tla.lir.oper.TlaTempOper
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.lir.transformations.TransformationTracker
+
+/**
+ * Encodes temporal formulas as invariants.
+ *
+ * @author
+ *   Philip Offtermatt
+ */
+@Singleton
+class TableauEncoder(
+    module: TlaModule,
+    gen: UniqueNameGenerator,
+    loopEnc: LoopEncoder,
+    tracker: TransformationTracker)
+    extends LazyLogging {
+  val levelFinder = new TlaLevelFinder(module)
+  val varNamesToExStrings = new HashMap[String, String]()
+
+  /**
+   * Encodes each of a sequence of temporal formulas.
+   * @see
+   *   [[encodeFormula]]
+   */
+  def temporalsToInvariants(
+      modWithPreds: ModWithPreds,
+      formulas: Seq[TlaOperDecl]): ModWithPreds = {
+
+    formulas.foldLeft(modWithPreds)(singleTemporalToInvariant)
```

Comment:
Because there's quite a bit of indirection it's hard to summarize as a single `suggestion`, but it's the same sentiment as before: instead of iteratively updating `mod` in a loop, it would make more sense, at least to me, to compute collections of all additions/overrides (e.g. separate collections of `Init_ex = ex` to be added to Init, and a `UNCHANGED << Init_ex1, ..., Init_ex2 >>` to be added to Next, as well as the implications for the formulas) and construct the final module once at the end.